Improved show for DataFrames

bkamins opened the pull request:
A proposal to solve #760.

Summary of changes:
- `show` and `showall` get a new argument `onechunk` which limits the number of printed chunks to 1 if `splitchunks` is `true` (with appropriate message in the summary)
- `showcols` gets two parameters `allcols` (if all columns of an `AbstractDataFrame` should be printed or only those fitting on the screen) and `values` (if sample of values of an `AbstractDataFrame` should be printed)
- `showcompact` is defined for `AbstractDataFrame` and `GroupedDataFrame`

nalimilan commented:
Thanks. These positional arguments are really getting out of hand. We should probably get rid of `splitchunks` (whose name is really confusing) and tell people to use `showcols` instead. Then `onechunk` could be renamed to `allcols`. What do you think?

I don't think `showcompact` is intended for this kind of use: as the docs say, it's mainly for scalar values to provide a short representation without type information to be used inside arrays.

nalimilan commented on src/abstractdataframe/show.jl:
```diff
@@ -556,32 +574,77 @@ end
 #' df = DataFrame(A = 1:3, B = ["x", "y", "z"])
-#' showcols(df, true)
-function showcols(io::IO, df::AbstractDataFrame) # -> Void
+#' showcols(STDOUT, df)
```
The example was correct, `STDOUT` is implicit.

bkamins replied:
I thould it should stay as the example with implicit `STDOUT` is given below for definition of `function showcols(df::AbstractDataFrame, allcols::Bool=false, values::Bool=true)`.

nalimilan replied:
OK. Anyway until we turn these into real docstrings this is pretty abstract.

nalimilan commented on src/abstractdataframe/show.jl:
```diff
@@ -556,32 +574,77 @@ end
+    nrows, ncols = size(df)
+    if values && nrows > 0
+        if nrows == 1
+            metadata[:Values] = [Symbol(df[1, i]) for i in 1:ncols]
```
Why a symbol? I think the best strategy is to use `sprint(showcompact, df[1, i])` (same below). In the tests, make sure to add cases with `String` and `Float32` columns, which are corner cases (the first needs quotes to be included inside the string, the other needs the `f0` prefix to be removed due to compact printing).

bkamins replied:
`sprint` would make `:Values` column a string and thus it would be rendered with additional enclosing double quotes when printed and symbol is printed "as is". Your suggestion is valuable though :) and I have modified the code.

nalimilan replied:
OK. But still use it before converting to symbol, to be sure you get the compact form. There have been many issues with this with the recent changes in 0.5.

nalimilan commented on src/abstractdataframe/show.jl:
```diff
@@ -556,32 +574,77 @@ end
 #' @param io::IO The `io` to be rendered to.
 #' @param df::AbstractDataFrame An AbstractDataFrame.
+#' @param allcols::Bool If `false` (default) only a subset of columns
+#'        fitting on the screen is printed
```
MIssing ending dot here and in other docstrings. Could also use a comma after "(default)" here and elsewhere.

bkamins commented:
@nalimilan Thx for the comments. I will correct the PR.

For the record let me add that this kind of output also should be fixed:
```
df = DataFrame(x=["a", "\t", "\\", "\n", "\$", "z"], y=1:6)
6×2 DataFrames.DataFrame
│ Row │ x   │ y │
├─────┼─────┼───┤
│ 1   │ "a" │ 1 │
│ 2   │ "\t"  │ 2 │
│ 3   │ "\\" │ 3 │
│ 4   │ "\n"  │ 4 │
│ 5   │ "\$" │ 5 │
│ 6   │ "z" │ 6 │
```

bkamins commented:
@nalimilan I hope I have covered all your comments correctly.

I have left `splitchunks` internally to differentiate the behavior of `show` and `showall` (similarly to arrays).

Additionally I have changed the formula calculating the width of the string so that `DataFrames` render correctly with escaped strings.

quinnj commented:
I think this would be a great fix; care to rebase?

bkamins commented:
Sure - I thought it was rejected.

quinnj commented:
Sorry, the package has struggled w/ maintenance over the last year or so as development moved over to DataTables.jl, but all that work has now been backported here and all development will now resume here.

nalimilan commented:
Sorry, I think I didn't finish the review because I wanted to get a complete understanding of the design space, and I didn't find the time for that at that point.

bkamins commented:
I have started to update the PR and have one issue to clarify.
There is a change how DataFrames performs `show` between latest release and master.

Consider running the following line in REPL:
```
df = DataFrame(A = 1:4,
               B = ["x\"", "y\n", "z\$", "ABC"],
               C = Float32[1.0, 2.0, 3.0, 4.0],
               D = Symbol[:ABC,Symbol("x\""),Symbol("y\n"),Symbol("z\$")])
```
on master it shows:
```
4×4 DataFrames.DataFrame
│ Row │ A │ B   │ C   │ D   │
├─────┼───┼─────┼─────┼─────┤
│ 1   │ 1 │ x"  │ 1.0 │ ABC │
│ 2   │ 2 │ y
   │ 2.0 │ x"  │
│ 3   │ 3 │ z$  │ 3.0 │ y
  │
│ 4   │ 4 │ ABC │ 4.0 │ z$  │
```
and latest release it shows:
```
4×4 DataFrames.DataFrame
│ Row │ A │ B     │ C   │ D   │
├─────┼───┼───────┼─────┼─────┤
│ 1   │ 1 │ "x\""  │ 1.0 │ ABC │
│ 2   │ 2 │ "y\n"   │ 2.0 │ x"  │
│ 3   │ 3 │ "z\$"  │ 3.0 │ y
  │
│ 4   │ 4 │ "ABC" │ 4.0 │ z$  │
```
and if we cast `df` to an array we get yet another output (it could be reproduced by `show` of DataFrame with column names and pipes added - here I want to concentrate on how the field values are printed):
```
julia> convert(Matrix, df)
4×4 Array{Any,2}:
 1  "x\""  1.0  :ABC
 2  "y\n"  2.0  Symbol("x\"")
 3  "z\$"  3.0  Symbol("y\n")
 4  "ABC"  4.0  Symbol("z$")
```
Which is the preferred target printing style? Or maybe yet some other option?
I personally would feel comfortable with the third one (show values like `show` for arrays) as it would be consistent.

quinnj commented:
I definitely think we should be consistent w/ array printing (last example). Show strings as quoted strings, as well as symbol that way. I think that's the only real solution if we want to be able to show unicode + control characters/whitespace and maintain the correct column widths.

nalimilan commented:
Agreed, the `Array` output looks like a good reference. Though printing quotes around strings is a bit verbose, and we could get rid of it if we printed the column eltype in a header (like [tibbles](https://github.com/tidyverse/tibble/blob/master/README.md) in R).

bkamins commented:
Thank you for the comments. Regarding refactoring of show I have some additional thoughts:
* if we omit `"` in strings how do we visually distinguish `"NA"` string from true `NA` (in R this is a problem with `"<NA>"`)
* how do we handle columns of custom types (normal and `Nullable` - whatever they will be eventually called) - in particular when their representation is very long (should some truncation be applied); in particular current `show` has problem for calculation of width of such structures, e.g.:
```
julia>     df = DataFrame(A=[[1:25;],"sdf"])
2×1 DataFrames.DataFrame
│ Row │ A                                                                                           │
├─────┼─────────────────────────────────────────────────────────────────────────────────────────────┤
│ 1   │ [1, 2, 3, 4, 5, 6, 7, 8, 9, 10  …  16, 17, 18, 19, 20, 21, 22, 23, 24, 25] │
│ 2   │ "sdf"                                                                                       │
```

In short - the problem is complex so I will try to give it some thought and will open a separate issue and write down my recommendation. I will strip this PR from field width calculation changes and leave only improved display features.

nalimilan commented:
> if we omit " in strings how do we visually distinguish "NA" string from true NA (in R this is a problem with "<NA>")

Yes, that would be a problem unless we add a header with the eltype of each column. But we can keep the quotes for now and discuss that possibility later.

>    how do we handle columns of custom types (normal and Nullable - whatever they will be eventually called) - in particular when their representation is very long (should some truncation be applied); in particular current show has problem for calculation of width of such structures, e.g.:

Some truncation should probably be applied (I think you can do that by setting a property on `IOContext` now). But that can also be improved later, no need to fix everything in a single PR. It's not that common to have fields like that in DataFrames anyway.

nalimilan commented:
@bkamins I think you need to rebase on the latest master. If there are still unrelated commits, use `git rebase -i master` and remove them.

bkamins commented:
Agreed - that is why in this PR I have left only changes to `show` global behavior and left other changes for the future.

One question regarding git: I have made a merge not a rebase and now I can see the bad consequences that all the intermediate commits got included. Is there any simple and safe way to fix it?

nalimilan commented:
Hmm... I guess the easiest solution would be to start a new branch from master, cherry-pick your commits into it, and then force push to this branch using `git push --force bkamins :newshow`.

bkamins commented:
Just as a comment: I believe that the build failed on Julia latest is unrelated to this PR.

nalimilan commented on src/abstractdataframe/show.jl:
```diff
@@ -543,32 +549,68 @@ end
+#' count.
+#'
+#' @param df::AbstractDataFrame An AbstractDataFrame.
+#' @param allcols::Bool If `false` (default), only a subset of columns
```
Maybe just call this `all`, since "col" is already clear from the function's name?

nalimilan replied:
This still applies, right?

nalimilan commented on test/show.jl:
```diff
@@ -1,19 +1,40 @@
     show(io, df, true)
     showall(io, df)
-    showall(io, df, true)
+    showall(io, df, false)
```
Could you test the actual output? You can use triple-quoted strings for that.

bkamins commented:
@nalimilan Inline comments got removed so I reply here:
* changed `Symbol("Row")` to `:Row`
* `@param values` comment string corrected
* all lines are below 92 chars
* I have removed `Symbol` in `:Values` formatting, but if we change the way `DataFrame` columns containing strings are printed it might have to be revised
* `allcols` changed to `all` (but one has to remember it clashes with `all` function)
* added testing of actual output in tests

nalimilan commented on src/abstractdataframe/show.jl:
```diff
@@ -297,6 +297,8 @@ end
 #' @param splitchunks::Bool Should the printing of the AbstractDataFrame
 #'        be done in chunks? Defaults to `false`.
+#' @param allcols::Bool Should only one chunk be printed if printing in
+#'        chunks? Defaults to `false`.
```
Defaults to `false`.

bkamins replied:
changed to `true`

nalimilan commented on src/abstractdataframe/show.jl:
```diff
@@ -315,21 +317,30 @@ function showrows(io::IO,
 
     rowmaxwidth = maxwidths[ncols + 1]
     chunkbounds = getchunkbounds(maxwidths, splitchunks, displaysize(io)[2])
     nchunks = length(chunkbounds) - 1
```
Would be clearer to do `nchunks = allcols ? length(chunkbounds) - 1 : min(nchunks, 1)`.

bkamins replied:
changed (with a bit different code as `nchunks` is undefined before this line)

nalimilan commented on src/abstractdataframe/show.jl:
```diff
@@ -543,32 +549,70 @@ end
-    showall(io, metadata, true, Symbol("Col #"), false)
+    nrows, ncols = size(df)
+    if values && nrows > 0
+        # type of Values column is now String; it might need to be changed
```
I don't think this comment is needed: tests will (or should) catch this and people will figure out what needs to be changed anyway.

bkamins replied:
removed

nalimilan commented on src/abstractdataframe/show.jl:
```diff
@@ -543,32 +549,70 @@ end
+        # type of Values column is now String; it might need to be changed
+        # if the way strings are printed in data frames changes
+        if nrows == 1
+            metadata[:Values] = [sprint(showcompact, df[1, i]) for i in 1:ncols]
```
Also, what about using `ourshowcompact`?

bkamins replied:
Changed - but it creates problems in corner cases (described in TODO for `getmaxwidths`)

nalimilan commented on test/show.jl:
```diff
@@ -1,19 +1,103 @@
+    4×3 DataFrames.DataFrame
+    │ Row │ A │ B             │ C   │
+    ├─────┼───┼───────────────┼─────┤
+    │ 1   │ 1 │ x\"            │ 1.0 │
```
I suppose the fact that vertical lines are not aligned is a bug elsewhere? Then better leave a `TODO` somewhere to make it clear.

bkamins replied:
They are aligned when the string is printed, but `"` needs to be escaped in string literal which breaks alignment in the code.

nalimilan replied:
Ah, of course!

bkamins replied:
Actually there is a problem - in the next line `│ 2   │ 2 │ ∀ε⫺0: x+ε⫺x │ 2.0 │` which is not aligned properly and it is a TODO do be added for `getmaxwidths` function. Sorry for confusion

nalimilan commented on test/show.jl:
```diff
@@ -1,19 +1,103 @@
-    show(io, df, true)
-    showall(io, df)
-    showall(io, df, true)
+    show(io, df_big)
```
Could you also test the output of these functions (even if that's verbose, it's fine)? Else the case when there are too many rows won't be covered. You should probably pass a custom `IOContext` to control the size of the display. Also better define `df_big` here rather than above, where it isn't used.

nalimilan commented on test/show.jl:
```diff
@@ -1,19 +1,103 @@
-    df = DataFrame(A = 1:3, B = ["x", "y", "z"])
+    # In the future newline characte \n should be added to this test case
+    df = DataFrame(A = 1:4, B = ["x\"", "∀ε⫺0: x+ε⫺x", "z\$", "ABC"],
+                   C = Float32[1.0, 2.0, 3.0, 4.0])
```
Could you add a `null` value somewhere so that this is covered (unless it's done elsewhere already)?

bkamins replied:
It is already covered I believe in line:
```
df = DataFrame(Fish = ["Suzy", "Amir"], Mass = [1.5, null])
```
at the end of the file

nalimilan replied:
Yes, but `showcols` isn't tested there. Would be worth adding a test.

bkamins replied:
added `showcols` test

bkamins commented:
@nalimilan I hope I have managed to clean up everything.

nalimilan commented on test/show.jl:
```diff
@@ -41,27 +39,181 @@ module TestShow
     refstr = """
     4×3 DataFrames.DataFrame
     
```
Just a detail, but we probably don't need an empty line? That would be more consistent with the other format.

bkamins replied:
removed

nalimilan commented on test/show.jl:
```diff
@@ -41,27 +39,181 @@ module TestShow
+    │ 24  │ 0.762276 │ 0.755415 │
+    │ 25  │ 0.339081 │ 0.649056 │"""
+
+    io = IOContext(IOBuffer(), :displaysize=>(10,40))
```
Maybe set the number of rows to a lower value in order to have smaller test and check what happens when not all rows can be shown in a single page? Can also be done in a later PR if you prefer.

bkamins replied:
I believe this is what I check here. I assume 10 rows and 40 columns. And you can see the difference between `show` and `showall`. 
`show` limits the output to fit page height and `showall` does not do that.
They also differ in how they handle *wide* data (not fitting the screen vertically) and we set `allcols` to `true`: `show` does paging and `showall` prints full table ignoring `:displaysize` (which could useful, when e.g. we want to dump `DataFrame` show result to a file).

bkamins commented:
@cjprybol fixed the issue with tests on 32 bit machine.

nalimilan commented:
Thanks! Merging since Travis doesn't seem to be willing to run on Mac...
